### DISCUSSION TITLE
UniqueIntegrityError

### DISCUSSION THREAD

**Discussion by pmithrandir** [sqlalchemy/sqlalchemy#12098] [Ideas]
Hello,

In case a query break a unique constraint, an IntegrityError is raised.

My issue is that I get the same Integrity Error for other problem (Not Null constraint, etc...)

I don't think we can differentiate the error returned which prevent to recover smoothly.

For example, if you can't add an item which already exist, you might want to update it instead.(we can imagine plenty of reason to know what failed)

Would it be possible to raise new sort of exception, inheriting from IntegrityError for backward compatibility, but more specialized ?
UniqueConstraintIntegrityError
ForeignKeyConstrraintError
NotNullIntegrityError
....

Best regards,
Pierre

**Comment by zzzeek** [comment]
unfortunately no since these exceptions [originate from your database driver, not SQLAlchemy](https://docs.sqlalchemy.org/en/20/core/exceptions.html#sqlalchemy.exc.DBAPIError).  if you need to differentiate between different `IntegrityError` scenarios you'll need to either look for attributes like the string message and possibly `.code` if present

**Reply by pmithrandir** [reply]
thank you for that precision.

Do you imagine sqlalchemy, or the drivers, could make that conversion available for everyone ?

First by creating these ecxeption in sqlalchemy, then by having each driver capture the dbapi exception, parsing it and raising the more precise sqlalchemy error ?

**Reply by zzzeek** [reply]
This is a really old topic and I am sorry to say the subject is quite calcified in a situation that was never thought out well within the DBAPI itself.    Every new driver, new database, new version of that driver/database changes things around so a generalized solution that parses exception messages will always be a moving target and not something you can reasonably maintain at a generic "one size fits all" level.   For an example of a framework that parses exceptions based on various criteria see the work I did in [openstack oslo.db](https://github.com/openstack/oslo.db/blob/master/oslo_db/sqlalchemy/exc_filters.py)
